llvm/lib/Transforms/Utils/SCCPSolver.cpp

```diff
@@ -1534,8 +1535,10 @@ void SCCPInstVisitor::visitBinaryOperator(Instruction &I) {
     return markOverdefined(&I);
 
   // Try to simplify to a constant range.
-  ConstantRange A = getConstantRange(V1State, I.getType());
-  ConstantRange B = getConstantRange(V2State, I.getType());
+  ConstantRange A =
+      getConstantRange(V1State, I.getType(), /*UndefAllowed=*/false);
+  ConstantRange B =
```

Comment:
I think the code that uses the constant range information from ValueLatticeElements already should be handling the range-may-include-undef case properly (i.e. not use it when refining instructions, only when using it to replace with a constant).

But it looks like the code below (`mergeInValue` call) simply drops the may-include-undef bit  from the operands. If we propagate the bit to the result, things should be fine?

```
-  mergeInValue(&I, ValueLatticeElement::getRange(R));
+  mergeInValue(&I, ValueLatticeElement::getRange(
+                       R, V1State.isConstantRangeIncludingUndef() ||
+                              V2State.isConstantRangeIncludingUndef()));
```

(IIRC the may-inlcude-undef bit was explicitly added to account for cases such as the AND test cases and I was pretty sure this was covered, but perhaps this was a place that got missed)

Reply:
If the above works out, it may make sense to make the `MayIncludeUndef` argument for `ValueLatticeElement::getRange()` required or flip the default?

Reply:
Worth noting that the "treat undef as full range" approach is also what we do in LVI.

Reply:
Ah yes that makes sense, thanks!